llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -880,6 +880,8 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
       setOperationAction(Op, MVT::f16, Legal);
   }
   // clang-format on
+  if (!Subtarget->hasFullFP16())
+    setOperationPromotedToType(ISD::FCANONICALIZE, MVT::f16, MVT::f32);
```

Comment:
Can this be added to LegalizeNarrowFP? I think that f32 is the default promotion type for f16.